src/twisted/internet/test/test_inlinecb.py

```diff
@@ -145,3 +147,186 @@ def calling():
+    """
+    Sample exception type.
+    """
+
+    def __init__(self, actual):
+        Exception.__init__(self)
+        self.actualValue = actual
+
+
+
+class CancellationTests(SynchronousTestCase):
+    """
+    Tests for cancellation of L{Deferred}s returned by L{inlineCallbacks}.
+    For each of these tests, let:
+        - C{G} be a generator decorated with C{inlineCallbacks}
+        - C{D} be a L{Deferred} returned by C{G}
+        - C{C} be a L{Deferred} awaited by C{G} with C{yield}
+    """
+
+    def setUp(self):
+        """
+        Set up the list of outstanding L{Deferred}s.
+        """
+        self.deferredsOutstanding = []
+
+
+    def tearDown(self):
+        """
+        If any L{Deferred}s are still outstanding, fire them.
+        """
+        while self.deferredsOutstanding:
+            self.deferredGotten()
+
+
+    @inlineCallbacks
+    def sampleInlineCB(self, getChildDeferred=None):
+        """
+        Generator for testing cascade cancelling cases.
+
+        @param getChildDeferred: Some callable returning L{Deferred} that we
+            awaiting (with C{yield})
+        """
+        if getChildDeferred is None:
+            getChildDeferred = self.getDeferred
+        try:
+            x = yield getChildDeferred()
+        except UntranslatedError:
+            raise TranslatedError()
+        except DontFail as df:
+            x = df.actualValue - 2
+        returnValue(x + 1)
+
+
+    def getDeferred(self):
+        """
+        A sample function that returns a L{Deferred} that can be fired on
+        demand, by L{CancellationTests.deferredGotten}.
+
+        @return: L{Deferred} that can be fired on demand.
+        """
+        self.deferredsOutstanding.append(Deferred())
+        return self.deferredsOutstanding[-1]
+
+
+    def deferredGotten(self, result=None):
+        """
+        Fire the L{Deferred} returned from the least-recent call to
+        L{CancellationTests.getDeferred}.
+
+        @param result: result object to be used when firing the L{Deferred}.
+        """
+        self.deferredsOutstanding.pop(0).callback(result)
+
+
+    def test_cascadeCancellingOnCancel(self):
+        """
+        When C{D} cancelled, C{C} will be immediately cancelled too.
+        """
+        childResultHolder = ['FAILURE']
+        def getChildDeferred():
+            d = Deferred()
+            def _eb(result):
+                childResultHolder[0] = result.check(CancelledError)
+                return result
+            d.addErrback(_eb)
+            return d
+        d = self.sampleInlineCB(getChildDeferred=getChildDeferred)
+        d.addErrback(lambda result: None)
+        d.cancel()
+        self.assertEqual(
+            childResultHolder[0],
+            CancelledError,
+            "no cascade cancelling occurs",
+        )
+
+
+    def test_errbackCancelledErrorOnCancel(self):
+        """
+        When C{D} cancelled, CancelledError from cascade cancelled C{C} will be
+        trapped
```

Comment:
Nothing is trapped here; the `CancelledError` is sent from `D` to `C`, which doesn't do anything with it, so it's then sent back to `D`.